core/src/docs/rfcs/2779_list_with_metakey.md

```diff
@@ -0,0 +1,126 @@
+- Start Date: 2023-08-04
+- RFC PR: [apache/incubator-opendal#2779](https://github.com/apache/incubator-opendal/pull/2779)
+- Tracking Issue: [apache/incubator-opendal#0000](https://github.com/apache/incubator-opendal/issues/0000)
+
+# Summary
+
+Move `Operator` `metadata` API to `list_with().metakey()` to simplify the usage.
+
+# Motivation
+
+The current `Entry` metadata API is:
+
+```rust
+use opendal::Entry;
+use opendal::Metakey;
+
+let meta = op
+    .metadata(&entry, Metakey::ContentLength | Metakey::ContentType)
+    .await?;
+```
+
+This API is difficult to understand and rarely used correctly. And in reality, users always fetch the same set of metadata during listing.
+
+Take one of our users code as an example:
+
+```rust
+let stream = self
+    .inner
+    .scan(&path)
+    .await
+    .map_err(|err| format_object_store_error(err, &path))?;
+
+let stream = stream.then(|res| async {
+    let entry = res.map_err(|err| format_object_store_error(err, ""))?;
+    let meta = self
+        .inner
+        .metadata(&entry, Metakey::ContentLength | Metakey::LastModified)
+        .await
+        .map_err(|err| format_object_store_error(err, entry.path()))?;
+
+    Ok(format_object_meta(entry.path(), &meta))
+});
+
+Ok(stream.boxed())
+```
+
+By moving metadata to `lister`, our user code can be simplified to:
+
+```rust
+let stream = self
+    .inner
+    .scan_with(&path)
+    .metakey(Metakey::ContentLength | Metakey::LastModified)
+    .await
+    .map_err(|err| format_object_store_error(err, &path))?;
+
+let stream = stream.then(|res| async {
+    let entry = res.map_err(|err| format_object_store_error(err, ""))?;
+    let meta = entry.into_metadata()
+
+    Ok(format_object_meta(entry.path(), &meta))
+});
+
+Ok(stream.boxed())
+```
+
+By introducing this change:
+
+- Users don't need to capture `Operator` in the closure.
+- Users don't need to do async call like `metadata()` again.
+
+If we don't have this change:
+
+- every place that could receive a `fn()` must use `Fn()` instead which enforce users to have a generic parameter in their code.
+- It's harder for other languages binding to implement `op.metadata()` right.
+
+# Guide-level explanation
+
+The new API will be:
+
+```rust
+let entries: Vec<Entry> = op
+  .list_with("dir")
+  .metakey(Metakey::ContentLength | Metakey::ContentType).await?;
+
+let meta: &Metadata = entries[0].metadata();
+```
+
+Metadata can be queried directly when listing entries via `metadata()`, and later extracted via `into_parts()`.
+
+# Reference-level explanation
+
+We will add `metakey` into `OpList`. Underlying services can use those information to try their best to fetch the metadata.
+
+There are following possibilities:
+
+- The entry metadata is met: `Lister` return the entry directly
+- The entry metadata is not met and not fully filled: `Lister` will try to send `stat` call to fetch the metadata
+- The entry metadata is not met and fully filled: `Lister` will return the entry directly.
```

Comment:
If the metadata is not met but fully filled, why are we returning the entry? I thought the Metakey works somehow like a filter.

Reply:
Metakey is more like a request query. It's something like: Hi, I expect this entry to have `content_type` and `etag`.

However, it is possible that the service does not have this metadata at all. For example, `fs` doesn't have `etag`. We will set `etag` to `None` to indicate that this entry doesn't have `etag`, and it's the final result, don't need to call `stat` again.

Reply:
What is the difference between those three cases? Those people who have not used the `list`, may not know.

I suggest specifying the scenarios where an Entry will appear. 
Some services' `list` directly return metadata, such as [S3](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html). However, others do not, which requires them to make an additional `stat` to achieve the same output.

Reply:
What is the difference between those three cases? Those people who have not used the `list`, may not know.

I suggest specifying the scenarios where an Entry will appear. 
Some services' `list` directly return metadata, such as [S3](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html). However, others do not, which requires them to make an additional `stat` to achieve the same output.

Reply:
`Entry` is a compose of `Path` and `Metadata`. Every service will return `Entry` during list, the difference is that different service may return differnt set of metadata.

Reply:
cc @Ji-Xinyou, would you like to take another look?

Reply:
In general, the design LGTM. Whether certain type of Metadata is returned should be checked by the users. But I think for each service we may specify them through the documentation?

It could be the case that users want certain type of metadata, but the service does not support that at all(which leads to returning None). I don't think case like this can be avoided on the code level, so the users have to read the storage documentation.

WDYT

Reply:
> But I think for each service we may specify them through the documentation?

Maybe so, but I'm afraid that specifying them in the documentation would not be helpful. It is already very clear that services do not have this metadata where users find the value to be `None`.

Anyway, it's good to be included in documents as an addition. But I will not include them in this RFC.

Reply:
I agree with you. The RFC is already complete enough.

Nit: I think this API will need careful documentation. Since we guarantees giving the Metakey only when the service supports that. (Maybe we can include this in the rfc? Because when I read the three cases of the list_with, it takes a while to comprehend that)

Reply:
> Since we guarantees giving the Metakey only when the service supports that.

No, there is something you have misunderstood. The restriction is for OpenDAL instead for users.

Let's take some real examples:

`fs` service only have `content_length` and `last_modified` metadata. But it's valid for users try to query metakey like `content_length | last_modified | content_type`.

OpenDAL will check the user's metakey with the metadata return by `fs` service. OpenDAL found that `fs` only has `content_length` and `last_modified` metadata, so it will set `content_type` to `None` to indicate users that the file doesn't have `content_type` metadata at all.

The only thing users need to do is check `meta.content_type().is_some()` before using it.

Again: It's valid for users to query any metadata they want. OpenDAL will make sure that the metadata they will be valid state: `Some(T)` or `None`. And users don't need to try `stat` again.

Reply:
I got your point now😘

Reply:
I will add some context for how `metakey` works in RFC.